File: 1840/solution.java
class Solution {
    public int maxBuilding(int n, int[][] restrictions) {
        List<int[]> restrict = new ArrayList<>();
        for(int[] arr: restrictions)
            restrict.add(arr);
        restrict.add(new int[]{1,0});
        Collections.sort(restrict, (a,b)-> a[0]-b[0]);

        if(restrict.get(restrict.size()-1)[0]!=n)
            restrict.add(new int[]{n,n-1});
        
        int r = restrict.size();
        int res = 0, h1, h2, dist;
        for(int i=1;i<r;i++){
            dist = restrict.get(i)[0] - restrict.get(i-1)[0];
            restrict.get(i)[1] = Math.min(restrict.get(i)[1], restrict.get(i-1)[1]+dist);
        }

        for(int i=r-2;i>=0;i--){
            dist = restrict.get(i+1)[0] - restrict.get(i)[0];
            restrict.get(i)[1] = Math.min(restrict.get(i)[1], restrict.get(i+1)[1]+dist);
        }

        for(int i=1;i<r;i++){
            dist = restrict.get(i)[0] - restrict.get(i-1)[0];
            h1 = restrict.get(i-1)[1];
            h2 = restrict.get(i)[1];
            res = Math.max(res, (h1+h2+dist)/2);
        }

        return res;
    }
}

File: 1840/solution.py
class Solution:
    def maxBuilding(self, n: int, restrictions: List[List[int]]) -> int:
        restrictions.append([1,0])
        restrictions.sort()
        if restrictions[-1][0]!=n:
            restrictions.append([n,n-1])
        r = len(restrictions)
        for i in range(1,r):
            dist = restrictions[i][0] - restrictions[i-1][0]
            restrictions[i][1] = min(restrictions[i][1], restrictions[i-1][1]+dist)
        for i in range(r-2,-1,-1):
            dist = restrictions[i+1][0] - restrictions[i][0]
            restrictions[i][1] = min(restrictions[i][1], restrictions[i+1][1]+dist)
        res = 0 
        for i in range(1,r):
            dist = restrictions[i][0] - restrictions[i-1][0]
            h1,h2 = restrictions[i-1][1], restrictions[i][1]
            res = max(res, (h1+h2+dist)//2)
        return res
